src/graph.rs

```diff
@@ -223,3 +224,205 @@ pub fn compare_unit_props(
 
     return Ok(Some(result_if_found));
 }
+
+pub fn determine_if_included(
+    db: &Connection,
+    earlier_unit: &String,
+    later_units: &[&String],
+) -> Result<bool> {
+    if storage::is_genesis_unit(&earlier_unit) {
+        return Ok(true);
+    }
+
+    let (earlier_unit_props, later_units_props) =
+        storage::read_props_of_units(db, &earlier_unit, later_units)?;
+
+    if earlier_unit_props.is_free == 1 {
+        return Ok(false);
+    }
+
+    ensure!(later_units_props.len() > 0, "no later unit props were read");
+
+    //spec::UnitProps.latest_included_mc_index and spec::UnitProps.main_chain_index is not Option
+    let max_later_limci = later_units_props
+        .iter()
+        .max_by_key(|props| props.latest_included_mc_index)
+        .unwrap()
+        .latest_included_mc_index;
+    if
+    /*earlier_unit_props.main_chain_index.is_some()
+        &&*/
+    max_later_limci >= earlier_unit_props.main_chain_index {
+        return Ok(true);
+    }
+
+    let max_later_level = later_units_props
+        .iter()
+        .max_by_key(|props| props.level)
+        .unwrap()
+        .level;
+    if max_later_level < earlier_unit_props.level {
+        return Ok(false);
+    }
+
+    let mut start_units = later_units
+        .iter()
+        .map(|s| format!("'{}'", s))
+        .collect::<Vec<_>>();
+
+    'go_up: loop {
+        let start_unit_list = start_units
+            .iter()
+            .map(|s| format!("'{}'", s))
+            .collect::<Vec<_>>()
+            .join(", ");
```

Comment:
you construct start_unit_list from star_unit which seems wrong, should be construct from later_units, or you have wrongly initialized start unit.